database/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala

```diff
@@ -0,0 +1,144 @@
+                                   attempt: Int)
+                                  (implicit ec: ExecutionContext): Future[Seq[Metadatum]] = {
+    val action = dataAccess.metadataByWorkflowUuidAndKeyAndCallFqnAndIndexAndAttempt(workflowUuid, key, callFqn, index,
+      attempt).result
+    runTransaction(action)
+  }
+
+  override def queryMetadataEventsWithWildcardKeys(workflowUuid: String,
+                                                   wildcardKeys: NonEmptyList[String],
+                                                   requireEmptyJobKey: Boolean)
+                                                  (implicit ec: ExecutionContext): Future[Seq[Metadatum]] = {
+    val action = dataAccess.queryMetadataMatchingAnyWildcardKeys(workflowUuid, wildcardKeys, requireEmptyJobKey).result
+    runTransaction(action)
+  }
+
+  override def queryMetadataEventsWithoutWildcardKeys(workflowUuid: String,
+                                                      wildcardKeys: NonEmptyList[String],
+                                                      requireEmptyJobKey: Boolean)
+                                                     (implicit ec: ExecutionContext): Future[Seq[Metadatum]] = {
+    val action = dataAccess.queryMetadataNotMatchingAnyWildcardKeys(workflowUuid, wildcardKeys,
+      requireEmptyJobKey).result
+    runTransaction(action)
+  }
+
+  private def updateMetadata(buildUpdatedSummary:
+                             (Option[WorkflowMetadataSummary], Seq[Metadatum]) => WorkflowMetadataSummary)
+                            (metadataByUuid: (String, Seq[Metadatum]))(implicit ec: ExecutionContext): DBIO[Unit] = {
+    val (uuid, metadataForUuid) = metadataByUuid
+    for {
+    // There might not be a preexisting summary for a given UUID, so `headOption` the result
+      existingSummary <- dataAccess.workflowMetadataSummariesByUuid(uuid).result.headOption
+      updatedSummary = buildUpdatedSummary(existingSummary, metadataForUuid)
+      _ <- upsertWorkflowMetadataSummary(updatedSummary)
+    } yield ()
+  }
+
+  private def upsertWorkflowMetadataSummary(summary: WorkflowMetadataSummary)
+                                           (implicit ec: ExecutionContext): DBIO[Unit] = {
+    if (useSlickUpserts) {
+      for {
+        _ <- dataAccess.workflowMetadataSummaryAutoInc.insertOrUpdate(summary)
+      } yield ()
+    } else {
+      for {
+        updateCount <- dataAccess.workflowMetadataSummariesByUuid(summary.workflowUuid).update(summary)
+        _ <- updateCount match {
+          case 0 => dataAccess.workflowMetadataSummaryAutoInc += summary
+          case 1 => DBIO.successful(Unit)
+          case _ => DBIO.failed(new RuntimeException(s"Unexpected summary update count $updateCount"))
+        }
+      } yield ()
+    }
+  }
+
+  def refreshMetadataSummaries(startId: Long, startTimestamp: Option[Timestamp],
+                               key1: String, key2: String, key3: String, key4: String,
```

Comment:
Yep. The other choice is to just not used a compiled query.

Because refreshing happens on a fixed schedule, and not in sudden bursts, I don't think recompiling the slick query is that expensive, and we could switch to an `inSet` if we wanted.